vmaas/reposcan/katello.py

```diff
@@ -0,0 +1,181 @@
+    ) -> dict:
+        """Sends request to Katello server/API"""
+
+        if scheme == "https":
+            verify = os.path.join(self.tmp_directory, os.path.basename(KATELLO_CA_CERT_PATH))
+            if not os.path.isfile(verify):
+                LOGGER.error("Katello CA certificate not found! Skipping API call.")
+                self.success = False
+                return {}
+            auth = (self.api_user, self.api_pass)
+        else:
+            verify = True
+            auth = None
+
+        if json:
+            headers = {"Accept": "application/json"}
+        else:
+            headers = {}
+
+        url = f"{scheme}://{self.hostname}{endpoint}"
+        tries = 0
+
+        while True:
+            if tries >= RETRY_COUNT:
+                break
+            try:
+                response = requests.request(
+                    method="GET",
+                    url=url,
+                    headers=headers,
+                    timeout=timeout,
+                    auth=auth,
+                    verify=verify,
+                    **kwargs,
+                )
+                if response.status_code == 200:
+                    if json:
+                        return response.json()
+                    return {"data": response.text}
+
+                tries += 1
+                LOGGER.error("Error during GET request to url %s: HTTP %s, %s", url, response.status_code, response.text)
+                # Do not retry for 4xx HTTP codes
+                if 400 <= response.status_code < 500:
+                    break
+            except requests.exceptions.RequestException:
+                tries += 1
+                LOGGER.exception("Error calling API %s: ", url)
+        self.success = False
+        return {}
+
+    def _download_katello_ca_certificate(self) -> str | None:
+        LOGGER.info("Downloading Katello CA certificate.")
+
+        ca_certificate = self._fetch_katello(KATELLO_CA_CERT_PATH, scheme="http", json=False)
+        if not ca_certificate:
+            LOGGER.error("Katello CA certificate download failed.")
+            self.success = False
+            return None
+
+        # Write CA certificate to tempfile to be able to use with requests
+        with open(os.path.join(self.tmp_directory, os.path.basename(KATELLO_CA_CERT_PATH)), "w", encoding="utf-8") as ca_cert_tmpfile:
+            ca_cert_tmpfile.write(ca_certificate["data"])
+
+        return ca_certificate["data"]
+
+    def _download_katello_access_certificate(self, org_id: int, org_label: str) -> str | None:
+        LOGGER.info("Downloading Katello organization '%s' access certificate.", org_label)
+
+        access_certificate_key = self._fetch_katello(KATELLO_ACCESS_CERT_PATH % org_id, json=False)
+        if not access_certificate_key:
+            LOGGER.error("Katello access certificate download failed, skipping organization '%s'.", org_label)
+            self.success = False
```

Comment:
I'd suggest to replace the instance variable with an error that is being returned or an exception raised (with exception handling in caller(s).

Reply:
Changed it to use exception and return success bool only from the public method.